mundane-dsv/src/main/scala/com/ambiata/mundane/dsv/DelimitedValuesParser.scala

```diff
@@ -0,0 +1,68 @@
+package com.ambiata.mundane
+package dsv
+
+import scala.collection.JavaConversions._
+
+import org.apache.commons.csv.{CSVParser, CSVFormat}
+
+import scalaz._, Scalaz._
+
+/**
+ * Create a parser for delimited values based on
+ * different implementations and a choice of
+ *
+ *  - delimiter
+ *  - character used for quoting fields
+ *  - character used to escaped control characters (delimiter, quoteCharacter)
+ *
+ */
+object DelimitedValuesParser {
+
+  def createParser(delimiter: Char, quoteCharacter: Option[Char], escapeCharacter: Option[Char], implementation: DelimitedValuesParserImplementation): DelimitedValuesParser =
+    implementation match {
+      case CommonsCsv => CommonsDelimitedValuesParser(delimiter, quoteCharacter, escapeCharacter)
+    }
+
+  /** use the CommonsCsv implementation as the default implementation */
+  def delimited(delimiter: Char): DelimitedValuesParser =
+    createParser(delimiter, '"'.some, none[Char], CommonsCsv)
+
+  /** use the CommonsCsv implementation as the default implementation */
+  def csv: DelimitedValuesParser =
+    delimited(',')
+
+  /** use the CommonsCsv implementation as the default implementation */
+  def psv: DelimitedValuesParser =
+    delimited('|')
+
+  /** use the CommonsCsv implementation as the default implementation */
+  def tsv: DelimitedValuesParser =
+    delimited('\t')
+}
+
+sealed trait DelimitedValuesParser {
+  def parse(line: String): String \/ List[String]
+}
+
+/**
+ * Implementation of a DelimitedValuesParser based on the Apache Commons CSV library
+ */
+case class CommonsDelimitedValuesParser(delimiter: Char, quoteCharacter: Option[Char], escapeCharacter: Option[Char]) extends DelimitedValuesParser {
+  private[this] val format: CSVFormat = {
+    List(
+      (quoteCharacter, (f: CSVFormat, c: Char) => f withQuote c),
+      (escapeCharacter, (f: CSVFormat, c: Char) => f withEscape c)).foldLeft(CSVFormat.newFormat(delimiter)) {
+      case (res, (ch, function)) =>
+        ch.fold(res)(c => function(res, c))
+    }
+  }
+
+  def parse(line: String): String \/ List[String] =
+    \/.fromTryCatchNonFatal(
+      CSVParser.parse(line, format).getRecords.toList.headOption.map(_.iterator.toList).getOrElse(Nil)).leftMap(_.getMessage)
+}
+
+
+sealed trait DelimitedValuesParserImplementation
+
+object CommonsCsv extends DelimitedValuesParserImplementation
```

Comment:
I think we will have different implementations, at least until we are comfortable we actually have a super-set of csvish (if one actually exists). I agree that just having createParserXYZ makes a lot of sense, we wouldn't be writing code to support multiple different versions - but different pieces of code - would use different versions, and it is fine (necessary even - from a correctness / complexity management standpoint) for different pieces of code to only work with 1 implementation.